### PR TITLE
Fixes #9711: Add generic methods to share files between nodes using rudder relay API

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/9711

### CODE REVIEW

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#####################################################################################
+# Copyright 2013 Normation SAS
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: 2017

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile from node
+# @description This method retreive a file shared from another rudder node
+# @documentation This method eanotheve a file shared fromr rudder node using a unique file identifier.
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: eanotheve -> retreives ?

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#  The file will be downloaded using cfengine protocol and copied into a new file.
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: CFEngine

Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: Could you also add an example (to see what id look like, that the path is absolute, etc.)

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+#
+#####################################################################################
+
+# @name Sharedfile to node
+# @description This method share a file with another rudder node
+# @documentation This method share a file with another rudder node using a unique file identifier.
+#  The file will be kept on the policy server and transmitted to the destination node's policy server if it is different.
+#  It will be kept on this server for the destination node to download as long as it is not replaced by a new
+#  file with the same id or remove by expiration of the TTL. 
+#  The TTL cas be a number of second or a string of the long form "1day 2hours 3minute 4seconds" or abbreviated in the form "5h 3s"
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: cas -> can

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile to node
+# @description This method share a file with another rudder node
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: shares

Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: Rudder

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+# @parameter file_id      unique name that was used to identify the file on the sender
+# @parameter file_path    where to put the file content
+# 
+# @class_prefix sharedfile_from_node
+# @class_parameter file_id
+
+bundle agent sharedfile_from_node(source_uuid, file_id, file_path)
+{
+  vars:
+      "old_class_prefix"  string => canonify("file_ensure_lines_present_${file_id}");
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: file_ensure_lines_present -> sharedfile_from_node

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+# @parameter file_path    path of the file to share
+# @parameter ttl          time to keep the file on the policy server in seconds or in human readable form (see long description)
+# 
+# @class_prefix sharedfile_to_node
+# @class_parameter file_id
+
+bundle agent sharedfile_to_node(target_uuid, file_id, file_path, ttl)
+{
+  vars:
+      "old_class_prefix"  string => canonify("file_ensure_lines_present_${file_id}");
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: file_ensure_lines_present -> sharedfile_to_node

**Comment by peckpeck** [comment]
Commit modified

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile from node
+# @description This method retreives a file shared from another Rudder node
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: retreives

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+# @documentation This method retreiveœ a file shared fromr rudder node using a unique file identifier.
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: retreives

Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: from a Ruddes

**Comment by peckpeck** [comment]
Commit modified

File: tree/30_generic_methods/sharedfile_from_node.cf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile from node
+# @description This method retreives a file shared from another Rudder node
+# @documentation This method retreives a file shared from a Rudder node using a unique file identifier.
+#  The file will be downloaded using CFEngine protocol and copied into a new file.
+#  The destination path must be the complete absolute path of the destination file.
+#  Ex: sharedfile_from_node("2bf1afdc-6725-4d3d-96b8-9128d09d353c", "application.properties", "/opt/application/etc/application.properties")
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_from_node.cf]: you can use markdown backquotes for the code

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile to node
+# @description This method shares a file with another Rudder node
+# @documentation This method share a file with another rudder node using a unique file identifier.
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: shares, Rudder

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
Commit modified

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  vars:
+      "old_class_prefix"  string => canonify("sharedfile_to_node_${file_id}");
+      "promisers"          slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist  => { "${target_uuid}", "${file_id}", "${file_path}", "${ttl}" };
+
+      "canonified_path"   string => canonify("${file_path}");
+      "url"               string => "https://${def.policy_server}/rudder/relay-api/shared-files/${target_uuid}/${g.uuid}/${file_id}";
+      # TODO store this hash in the signature
+      "hash"              string => execresult("sha512sum ${file_path} | cut -d ' ' -f 1", "useshell");
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: This won't work on AIX.

Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: We may want to fail early with a clear error message when we can't find rudder-sign (vanilla CFEngine on Linux, Windows agents, etc.)

File: tree/30_generic_methods/sharedfile_to_node.cf [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+      # classes for _classes_combine_ifthenelse
+      "${old_class_prefix}_exist_true" => "${old_class_prefix}_exist_kept";
+      "${old_class_prefix}_exist_false" => "${old_class_prefix}_exist_repaired";
+      "${class_prefix}_exist_true" => "${class_prefix}_exist_kept";
+      "${class_prefix}_exist_false" => "${class_prefix}_exist_repaired";
+
+      "check_failed" => "${old_class_prefix}_exist_error|${class_prefix}_exist_error";
+
+  methods:
+    !_stdlib_path_exists_curl|check_failed::
```
Comment (amousset) [inline tree/30_generic_methods/sharedfile_to_node.cf]: We should split the "curl is not installed" and the "could not contact server" cases

**Comment by peckpeck** [comment]
Commit modified

**Pull request merged**